### PR TITLE
Tests for `Parser.__parse_type()`

### PR DESCRIPTION
**Pull request by magnopus-swifty**
Added fairly thorough tests for the `__parse_type()` function. These tests cover _almost_ every line of the function.

These tests build on each other, starting from checking simple types (e.g. `test_parse_primitive_type`), then using the parser to construct tests for templated types (`test_parse_template_type`) to finally checking templated `std::function<>` types with templated arguments.

`test_parse_pointer_to_pointer_to_pointer_type_fails` demonstrates that there are certain edge case types (`int***`) that the parser handles incorrectly.

### CODE REVIEW

File: Tools/WrapperGenerator/tests/test_parser.py [hunk 1]
```diff
@@ -78,3 +78,243 @@ def test_parse_enum_field_values(self):
...
+        expected = MetadataTypes.TypeMetadata("", "int")
+        expected.is_reference = True
+        expected.is_pointer_or_reference = True
+        expected.is_primitive = True
+        expected.template_name = "int"
+        expected.template_arguments = []
+
+        self.assertDictEqual(result.__dict__, expected.__dict__)
+
+    def test_parse_const_pointer_type(self):
```
Comment (MAG-mv) [inline Tools/WrapperGenerator/tests/test_parser.py:174]: [Consider] What about const int&

File: Tools/WrapperGenerator/tests/test_parser.py [hunk 2]
```diff
@@ -78,3 +78,243 @@ def test_parse_enum_field_values(self):
...
+        """ Test parsing a const pointer type. """
+        result = self.__parse_type("const int*")
```
Comment (MAG-mv) [inline Tools/WrapperGenerator/tests/test_parser.py:176]: [Question] I wonder if we support int const *?

File: Tools/WrapperGenerator/tests/test_parser.py [hunk 3]
```diff
@@ -78,3 +78,243 @@ def test_parse_enum_field_values(self):
...
+            MetadataTypes.TemplateArgumentMetadata(
+                type=int_type,
+                is_last=True
+            )
+        ]
+
+        self.assertDictEqual(result.__dict__, expected.__dict__)
+
+    def test_parse_multi_template_type(self):
+        """ Test parsing a template type with two arguments to the template. """
```
Comment (MAG-mv) [inline Tools/WrapperGenerator/tests/test_parser.py:226]: [Question] Is multiple template args actually fully supported? Or do we have a special case for maps?

Comment (MAG-mv) [reply Tools/WrapperGenerator/tests/test_parser.py:226]: Asking because we have a lot of special cases for certain types, although this may be in the language generation step

File: Tools/WrapperGenerator/tests/test_parser.py [hunk 4]
```diff
@@ -78,3 +78,243 @@ def test_parse_enum_field_values(self):
...
+        expected = MetadataTypes.TypeMetadata("std", "function")
+        expected.is_function_signature = True
+        expected.function_signature = function_metadata
+        expected.template_arguments = []
+        expected.template_name = "function"
+
+        self.assertDictEqual(result.__dict__, expected.__dict__)
+
+    def test_parse_function_with_return_type(self):
+        """ Test parsing a function type with a return type. """
```
Comment (MAG-mv) [inline Tools/WrapperGenerator/tests/test_parser.py:327]: [Consider] What about CSP macros? CSP_ASYNC_RESULT, CSP_EVENT, CSP_START_IGNORE etc

**Comment by MAG-mv** [comment]
[Comment] It looks like our parser seems well supported. A lot of our issues and limitations seem to come from the language generation step. Which makes sense, I guess.

File: Tools/WrapperGenerator/tests/test_parser.py (hunk 3 above)
Comment (magnopus-swifty) [reply Tools/WrapperGenerator/tests/test_parser.py:226]: I think the wrapper generator does (in theory) support it, at least for C#. The other tests (#779) were partly motivated by needing a place to hand-write tests that verify this.

File: Tools/WrapperGenerator/tests/test_parser.py (hunk 1 above)
Comment (magnopus-swifty) [reply Tools/WrapperGenerator/tests/test_parser.py:174]: See 1b4138c

File: Tools/WrapperGenerator/tests/test_parser.py (hunk 2 above)
Comment (magnopus-swifty) [reply Tools/WrapperGenerator/tests/test_parser.py:176]: Not right now. See 1b4138c.

File: Tools/WrapperGenerator/tests/test_parser.py (hunk 4 above)
Comment (magnopus-swifty) [reply Tools/WrapperGenerator/tests/test_parser.py:327]: I think I'd rather do a follow-up PR for testing `__parse_function` where I test function types specifically. I'll leave these tests here just to demonstrate how function metadata ends up in a type declaration but expect a future PR containing a `TestParserParseFunction` fixture.

**Review by MAG-AdamThorn — APPROVED** [review]
